src/hash.cr

```diff
@@ -357,6 +357,38 @@ class Hash(K, V)
     KeyIterator(K, V).new(self, @first)
   end
 
+  # Calls the given block for each matching value and passes in the key.
+  #
+  # ```
+  # h = {"foo" => "bar", "baz" => "qux"}
+  # h.each_key_for("qux") do |key|
+  #   key # => "baz"
+  # end
+  # # Note that "foo" is not included.
+  # ```
+  def each_key_for(value)
+    each do |k, v|
+      yield k if v == value
+    end
+  end
+
+  # Returns an iterator over the keys with a value matching the input value.
+  # Which behaves like an `Iterator` consisting of the key's types.
+  #
+  # ```
+  # h = {"one" => 1, "two" => 2, "dos" => 2}
+  # iterator = h.each_key_for(2)
+  #
+  # key = iterator.next
+  # key # => "two"
+  #
+  # key = iterator.next
+  # key # => "dos"
+  # ```
+  def each_key_for(value)
+    each.select { |(k, v)| v == value }.map &.first
```

Comment:
You should build a custom Iterator class for this. For example look at the valueiterator below.

Reply:
i thought this first, but what need? this functionality not very needed, so if not super optimized fine too

Reply:
@RX14 and @larubujo 
For moving to a custom Iterator class, what advantages does it provide?
- Better performance
- More reliable
- More readable
- More reusable

It will be very similar to `KeyIterator(K, V)`, from what I can tell.

`Entry#fore` may need to be overridden, unless there is a way to write `KeyForValueIterator#next` that I haven't figured out yet.

Reply:
@marksiemers more performance and more consistent with how the other iterators are implemented.

It wouldn't be hard, given the `BaseIterator` module and how the other iterators are implemented. It should in fact be only 3-4 lines in the `def next` impl.

Reply:
Actually my bad, it can't be done with `base_next`.

```cr
def next
  current = @current
  while current
    if current.value == @value
      @current = current.fore
      return current.key
    else
      current = current.fore
    end
  end
  @current = current
  stop
end
```